gpiozero/devices.py

```diff
@@ -229,13 +189,100 @@ def __exit__(self, exc_type, exc_value, exc_tb):
+
+    def _reserve_pins(self, *pins_or_addresses):
+        """
+        Called to indicate that the device reserves the right to use the
+        specified *pins_or_addresses*. This should be done during device
+        construction.  If pins are reserved, you must ensure that the
+        reservation is released by eventually called :meth:`_release_pins`.
+
+        The *pins_or_addresses* can be actual :class:`Pin` instances or the
+        addresses of pin instances (each address is a tuple of strings). The
+        latter form is permitted to ensure that devices do not have to
+        construct :class:`Pin` objects to reserve pins. This is important as
+        constructing a pin often configures it (e.g. as an input) which
+        conflicts with alternate pin functions like SPI.
+        """
+        addresses = (
+            p.address if isinstance(p, Pin) else p
+            for p in pins_or_addresses
+            )
+        with self._res_lock:
+            for address in addresses:
+                try:
+                    conflictors = self._reservations[address]
+                except KeyError:
+                    conflictors = []
+                    self._reservations[address] = conflictors
+                for device_ref in conflictors:
+                    device = device_ref()
+                    if device is not None and self._conflicts_with(device):
+                        raise GPIOPinInUse(
+                            'pin %s is already in use by %r' % (
+                                '/'.join(address), device)
+                        )
+                conflictors.append(weakref.ref(self))
+
+    def _release_pins(self, *pins_or_addresses):
+        """
+        Releases the reservation of this device against *pins_or_addresses*.
+        This is typically called during :meth:`close` to clean up reservations
+        taken during construction. Releasing a reservation that is not
+        currently held will be silently ignored (to permit clean-up after
+        failed / partial construction).
+        """
+        addresses = (
+            p.address if isinstance(p, Pin) else p
+            for p in pins_or_addresses
+            )
+        with self._res_lock:
+            for address in addresses:
+                self._reservations[address] = [
+                    ref for ref in self._reservations[address]
+                    if ref() not in (self, None) # may as well clean up dead refs
+                    ]
+
+    def _release_all(self):
+        """
+        Releases all pin reservations taken out by this device. See
+        :meth:`_release_pins` for further information).
+        """
+        with self._res_lock:
+            Device._reservations = {
+                address: [
+                    ref for ref in conflictors
+                    if ref() not in (self, None)
+                    ]
+                for address, conflictors in self._reservations.items()
+                }
```

Comment:
I think perhaps bracket closing indentation and newlines? Personally I favour (a) but if you read PEP-8's [indentation section](https://www.python.org/dev/peps/pep-0008/#indentation) it's actually pretty ambivalent about this. Frankly, my only reason for preferring (a) is simply that it's vim's default so it's the path of least resistance for me, but I don't really care what people use so long as it's readable and vaguely PEP-8 compliant (I think Ben favours (c) for example).